src/ipa_pc/data_structures.rs

```diff
@@ -1,48 +1,104 @@
+// #[cfg(test)]
+use rand::thread_rng;
+// #[cfg(test)]
+// use algebra::UniformRand;
+
+/// `Parameters` are the polynomial commitment parameters for the inner product arg scheme.
+// TODO: it seems to be artificial to have a `Parameters` struct which is the same as
+// the committer key. Let us investigate if this struct is really needed.
+// (Can't we simply implement the PCParameters trait for the struct PCCommitterKey?)
 #[derive(Derivative)]
 #[derivative(
-    Default(bound = ""),
     Clone(bound = ""),
     Debug(bound = ""),
     Eq(bound = ""),
     PartialEq(bound = "")
 )]
 #[derive(CanonicalSerialize, CanonicalDeserialize)]
-pub struct UniversalParams<G: AffineCurve> {
+pub struct Parameters<G: Curve> {
     /// The key used to commit to polynomials.
-    pub comm_key: Vec<G>,
+    pub comm_key: Vec<G::AffineRep>,
 
-    /// Some group generator.
+    /// The base point that carries the inner products evaluations.
+    /// Used by the prover and verifier.
     pub h: G,
 
-    /// Some group generator specifically used for hiding.
+    /// The base point used for hiding.
     pub s: G,
 
-    /// H(comm_key, h, s, max_degree)
+    /// The finger print  H(comm_key, h, s, max_degree)
     pub hash: Vec<u8>,
 }
 
-impl<G: AffineCurve> PCUniversalParams for UniversalParams<G> {
+impl<G: Curve> PCParameters<G> for Parameters<G> {
+    type CommitterKey = CommitterKey<G>;
+    type VerifierKey = VerifierKey<G>;
+    type Error = Error;
+
     fn max_degree(&self) -> usize {
         self.comm_key.len() - 1
     }
+
     fn get_hash(&self) -> &[u8] {
         self.hash.as_slice()
     }
-    fn copy_params(&mut self, other: &Self) {
-        self.s = other.s;
-        self.h = other.h;
+
+    /// Trims the base point vector of the setup function to a custom segment size
+    fn trim(
+        &self,
+        segment_size: usize,
+    ) -> Result<(Self::CommitterKey, Self::VerifierKey), Self::Error> {
+        // Ensure that supported_degree + 1 is a power of two
+        let supported_degree = (segment_size + 1).next_power_of_two() - 1;
+        if supported_degree > self.max_degree() {
+            return Err(Error::TrimmingDegreeTooLarge);
+        }
+
+        let trim_time =
+            start_timer!(|| format!("Trimming to supported degree of {}", supported_degree));
+
+        let ck = CommitterKey {
+            comm_key: self.comm_key[0..(supported_degree + 1)].to_vec(),
+            h: self.h.clone(),
+            s: self.s.clone(),
+            max_degree: self.max_degree(),
+            hash: self.hash.clone(),
+        };
+
+        let vk = VerifierKey {
+            comm_key: self.comm_key[0..(supported_degree + 1)].to_vec(),
+            h: self.h.clone(),
+            s: self.s.clone(),
+            max_degree: self.max_degree(),
+            hash: self.hash.clone(),
+        };
+
+        end_timer!(trim_time);
+
+        Ok((ck, vk))
+    }
+
+    // #[cfg(test)]
+    fn ut_copy_params(&mut self, other: &Self) {
```

Comment:
Same comment as other ones

Reply:
Answered above

Reply:
Either you move these changes here and report them on ginger-lib monorepo branch,
or we formally decide to address this changes in ginger-lib, but you need to create a separate branch from the monorepo branch containing them.
Currently, it's difficult to review them as in the monorepo branch everything will appear as new.

Reply:
Done